src/sparsebase/io/mtx_writer.cc

```diff
@@ -0,0 +1,390 @@
+      }
+
+      //Symmetry check
+      bool saidSymmetric = (symmetry_ == "symmetric" || symmetry_ == "skew-symmetric" || symmetry_ == "hermitian");
+      int NNZ = coo->get_num_nnz();
+      int count_symmetric = 0;
+      int count_diagonal = 0;
+      int is_diagonal_all_zero = true;
+      if (saidSymmetric && dimensions[0] == dimensions[1]) 
+      {
+          for (int i = 0; i < coo->get_num_nnz(); ++i) 
+          {
+            if (row[i] != col[i]) 
+            { // Non-diagonal entry, check symmetric counterpart
+                bool found_symmetric = false;
+                for (int j = 0; j < coo->get_num_nnz(); ++j) 
+                {
+                  if (symmetry_ == "skew-symmetric")
+                  {
+                    if constexpr (std::is_same_v<ValueType, void>)
+                    {
+                      break; //pattern cannot be skew-symmetric
+                    }
+                    else
+                    {
+                      if (row[j] == col[i] && col[j] == row[i] && val[j] == -val[i]) 
+                      {
+                        found_symmetric = true;
+                        count_symmetric++;
+                        break;
+                      }
+                    }
+                  }
+                  else
+                  {
+                    if constexpr (std::is_same_v<ValueType, void>)
+                    {
+                      if (row[j] == col[i] && col[j] == row[i]) 
+                      {
+                        found_symmetric = true;
+                        count_symmetric++;
+                        break;
+                      }
+                    }
+                    else
+                    {
+                      if (row[j] == col[i] && col[j] == row[i] && val[j] == val[i]) 
+                      {
+                        found_symmetric = true;
+                        count_symmetric++;
+                        break;
+                      }
+                    }
+                  }
+                }
+                if (!found_symmetric) {
+                    throw utils::WriterException("Matrix is not symmetric!");
+                }
+            }
+            else //diagonal
+            {
+              count_diagonal++;
+              if constexpr (std::is_same_v<ValueType, void>)
+              {
+                //zero
+              }
+              else
+              {
+                if (val[i] != 0)
+                {
+                  is_diagonal_all_zero = false;
+                }
+              }
+            }
+          }
+          if(symmetry_ == "skew-symmetric")
+          {
+            if (!is_diagonal_all_zero)
+            {
+              throw utils::WriterException("Skew-symmetric matrix with non-zero diagonal values!");
+            }
+            NNZ -= (count_symmetric/2) + count_diagonal;
+          }
+          else
+          {
+            NNZ -= (count_symmetric/2);
```

Comment:
I think this should include the `+ count_diagonal` term, right? Since `symmetric` will have an element for each non-diagonal pair + an element for each element on the diagonal.